### DISCUSSION TITLE
Performance issue on large input

### DISCUSSION THREAD

**Discussion by kamyabzad** [davatorium/rofi#1620] [Q&A]
When using Rofi dmenu mode on large input (a couple of million lines), it's too slow, while dmenu itself is quite snappy. Does anybody have any idea how could I improve the performance?

Rofi command:
`$ cat dict.index | rofi -dmenu -sync`

dmenu command:
`$ cat dict.index | dmenu`

**Comment by DaveDavenport** [comment]
Rofi was not really designed to filter millions of lines. 

Can you explain what is slow? (opening, filtering, scrolling, displaying?)

**Reply by kamyabzad** [reply]
Opening time is slow but reasonable. I think it's about filtering, cause typing is quite laggy.

**Reply by DaveDavenport** [reply]
I can imagine as it uses regex to do matching and is utf-8 aware, that does not help. (It does use multi-threading for matching if you have not disabled this.)

Might be able to speed it up by tweaking some options, but I don´t expect wonders.
`-no-tokenize`  might help and possibly changing the matching algorithm.


For me loading 15mil lines takes 5.6 seconds with rofi (sync), and  3.8 with dmenu. That is reasonable.

**Reply by kamyabzad** [reply]
It will be great if we could improve filtering performance. Might tackle it in my free time. Thanks!

**Reply by DaveDavenport** [reply]
made some tweaks on the branch huge-list-speedups, this might improve things a bit by reducing the number of filter actions.

**Reply by DaveDavenport** [reply]
If possible please test that branch, I also rewrote the async reading (not to use GInputStream from glib) so that is now a lot! faster too.

**Reply by DaveDavenport** [reply]
reading 15 mil lines sync is now 3.8s for -sync, around ~5 seconds in async mode (to finish reading, its usable instantly).

**Reply by kamyabzad** [reply]
This is exactly what I had in mind! Thank you.
